File: eeg-asd-dashboard/core/feature_extraction.py
"""
Python port of feature_extraction_v2.m
=======================================
Extracts 79 features per subject:
  A) Spectral Power (Welch PSD): 42 features
     - Absolute & relative band power (5 bands × 4 regions × 2) = 40
     - Frontal theta/alpha and theta/beta ratios = 2
  B) Functional Connectivity (Coherence): 32 features
     - 8 electrode pairs × 4 bands
  C) Sample Entropy: 5 features
     - 4 regions + global

References:
- Original MATLAB: feature_extraction_v2.m
- Welch PSD: scipy.signal.welch
- Coherence: scipy.signal.coherence
- Sample Entropy: antropy.sample_entropy
"""

import numpy as np
from scipy.signal import welch, coherence
from typing import Dict, List, Tuple
import antropy


class EEGFeatureExtractor:
    """
    Extracts multi-domain EEG features matching the thesis methodology.
    
    Feature domains:
    1. Spectral Power (Welch PSD)
    2. Functional Connectivity (Magnitude-Squared Coherence)
    3. Nonlinear Complexity (Sample Entropy)
    """
    
    def __init__(self, srate: float = 256):
        """
        Configure feature extraction constants and build feature names.

        The bands, regions, coherence pairs, and entropy settings are kept in
        one place so every extracted feature follows the thesis ordering.
        """
        self.srate = srate
        
        # Frequency bands (matching thesis Table 4)
        self.bands = {
            'delta': (1, 4),
            'theta': (4, 8),
            'alpha': (8, 15),
            'beta': (15, 30),
            'gamma': (30, 40)
        }
        
        # Region definitions (matching MATLAB regionDefs)
        self.regions = {
            'frontal': ['Fp1','AF7','AF3','F1','F3','F5','F7',
                       'Fpz','Fp2','AF8','Fz','F4','F6','F8'],
            'central': ['FC5','FC3','FC1','FC4','FC2','C3','C5','C4','C6'],
            'parietal': ['P1','P3','P5','P7','P9','Pz','P2','P4','P6',
                        'P8','P10','CP3','CP4','CP6'],
            'occipital': ['PO7','PO3','POz','PO8','PO4','O1','Oz','O2','Iz']
        }
        
        # Coherence pairs (matching MATLAB connPairs)
        self.coherence_pairs = [
            ('F3', 'F4'),    # frontal interhemispheric short-range
            ('C3', 'C4'),    # central interhemispheric short-range
            ('P5', 'P6'),    # parietal interhemispheric (P5-P6 in GWO) short-range
            ('TP7', 'TP8'),  # temporal interhemispheric long-range
            ('O1', 'O2'),    # occipital interhemispheric short-range
            ('Fz', 'Pz'),   # midline fronto-posterior long-range   
            ('F3', 'P5'),    # left fronto-posterior long-range
            ('F4', 'P4'),    # right fronto-posterior long-range
        ]
        
        # Coherence bands (4 bands, no gamma for connectivity)
        self.conn_bands = {
            'delta': (1, 4),
            'theta': (4, 8),
            'alpha': (8, 15),
            'beta': (15, 30)
        }
        
        # Sample Entropy parameters (matching MATLAB)
        self.sampen_m = 2      # embedding dimension
        self.sampen_r = 0.2    # tolerance factor (r * std)
        
        # Build feature names
        self.feature_names = self._build_feature_names()
    
    def _build_feature_names(self) -> List[str]:
        """
        Build the ordered list of all 79 feature names.

        The order must match the numeric feature vector produced by extract()
        and the CSV/model artifacts used during inference.
        """
        names = []
        
        # A) Spectral: abs + rel per region per band (40 features)
        for region in self.regions:
            for band in self.bands:
                names.append(f'abs_{band}_{region}')
                names.append(f'rel_{band}_{region}')
        
        # Frontal ratios (2 features)
        names.append('theta_alpha_F')
        names.append('theta_beta_F')
        
        # B) Coherence (32 features: 8 pairs × 4 bands)
        pair_labels = ['F3_F4', 'C3_C4', 'P5_P6', 'TP7_TP8',
                       'O1_O2', 'Fz_Pz', 'F3_P5', 'F4_P4']
        for pair_label in pair_labels:
            for band in self.conn_bands:
                names.append(f'coh_{band}_{pair_label}')
        
        # C) Sample Entropy (5 features)
        for region in self.regions:
            names.append(f'sampen_{region}')
        names.append('sampen_global')
        
        return names
    
    def extract(self, epochs_data: np.ndarray, 
                channel_names: List[str]) -> np.ndarray:
        """
        Extract all 79 features from epoched EEG data.
        
        Parameters
        ----------
        epochs_data : np.ndarray
            Shape: [n_epochs, n_channels, n_samples]
        channel_names : list
            Channel labels matching the data
            
        Returns
        -------
        np.ndarray : Feature vector of length 79
        """
        n_epochs, n_channels, n_samples = epochs_data.shape
        
        # === A) SPECTRAL POWER (Welch PSD) ===
        # Compute PSD for each epoch and channel, then average over epochs
        win_len = min(self.srate, n_samples)  # 1-second window
        noverlap = win_len // 2
        
        # PSD: [n_epochs, n_channels, n_freqs]
        all_psd = []
        for ep in range(n_epochs):
            epoch_psd = []
            for ch in range(n_channels):
                sig = epochs_data[ep, ch, :]
                freqs, pxx = welch(sig, fs=self.srate, 
                                   nperseg=win_len, noverlap=noverlap)
                epoch_psd.append(pxx)
            all_psd.append(epoch_psd)
        
        all_psd = np.array(all_psd)  # [n_epochs, n_channels, n_freqs]
        mean_psd = np.mean(all_psd, axis=0)  # [n_channels, n_freqs]
        
        # Compute band powers per channel
        abs_power = {}  # {band: [n_channels]}
        for band_name, (f_low, f_high) in self.bands.items():
            band_mask = (freqs >= f_low) & (freqs <= f_high)
            if np.any(band_mask):
                abs_power[band_name] = np.trapz(
                    mean_psd[:, band_mask], freqs[band_mask], axis=1
                )
            else:
                abs_power[band_name] = np.zeros(n_channels)
        
        # Total power per channel
        total_power = sum(abs_power.values())
        total_power[total_power == 0] = np.nan
        
        # Relative power
        rel_power = {}
        for band_name in self.bands:
            rel_power[band_name] = abs_power[band_name] / total_power
        
        # Regional averages
        features = []
        for region_name, region_channels in self.regions.items():
            ch_indices = [i for i, ch in enumerate(channel_names) 
                         if ch in region_channels]
            
            for band_name in self.bands:
                if ch_indices:
                    abs_regional = np.nanmean(abs_power[band_name][ch_indices])
                    rel_regional = np.nanmean(rel_power[band_name][ch_indices])
                else:
                    abs_regional = np.nan
                    rel_regional = np.nan
                features.append(abs_regional)
                features.append(rel_regional)
        
        # Frontal theta/alpha and theta/beta ratios
        frontal_idx = [i for i, ch in enumerate(channel_names) 
                      if ch in self.regions['frontal']]
        if frontal_idx:
            theta_frontal = np.nanmean(rel_power['theta'][frontal_idx])
            alpha_frontal = np.nanmean(rel_power['alpha'][frontal_idx])
            beta_frontal = np.nanmean(rel_power['beta'][frontal_idx])
            
            theta_alpha_ratio = theta_frontal / alpha_frontal if alpha_frontal > 0 else np.nan
            theta_beta_ratio = theta_frontal / beta_frontal if beta_frontal > 0 else np.nan
        else:
            theta_alpha_ratio = np.nan
            theta_beta_ratio = np.nan
        
        features.append(theta_alpha_ratio)
        features.append(theta_beta_ratio)
        
        # === B) FUNCTIONAL CONNECTIVITY (Coherence) ===
        for ch1_name, ch2_name in self.coherence_pairs:
            idx1 = self._find_channel(ch1_name, channel_names)
            idx2 = self._find_channel(ch2_name, channel_names)
            
            if idx1 is not None and idx2 is not None:
                # Average coherence across epochs
                coh_epochs = []
                for ep in range(n_epochs):
                    f_coh, cxy = coherence(
                        epochs_data[ep, idx1, :],
                        epochs_data[ep, idx2, :],
                        fs=self.srate,
                        nperseg=win_len,
                        noverlap=noverlap
                    )
                    coh_epochs.append(cxy)
                
                mean_coh = np.mean(coh_epochs, axis=0)
                
                for band_name, (f_low, f_high) in self.conn_bands.items():
                    band_mask = (f_coh >= f_low) & (f_coh <= f_high)
                    if np.any(band_mask):
                        features.append(np.mean(mean_coh[band_mask]))
                    else:
                        features.append(np.nan)
            else:
                # Missing pair — append NaN for all 4 bands
                for _ in self.conn_bands:
                    features.append(np.nan)
        
        # === C) SAMPLE ENTROPY ===
        # Compute per channel, average across epochs first
        sampen_per_channel = np.zeros(n_channels)
        for ch in range(n_channels):
            ch_sampen = []
            for ep in range(n_epochs):
                sig = epochs_data[ep, ch, :]
                if np.std(sig) > 0:
                    try:
                        se = antropy.sample_entropy(
                            sig, order=self.sampen_m,
                            metric='chebyshev'
                        )
                        # antropy returns array; take last value
                        ch_sampen.append(se[-1] if hasattr(se, '__len__') else se)
                    except:
                        ch_sampen.append(np.nan)
                else:
                    ch_sampen.append(np.nan)
            sampen_per_channel[ch] = np.nanmean(ch_sampen)
        
        # Regional entropy
        for region_name, region_channels in self.regions.items():
            ch_indices = [i for i, ch in enumerate(channel_names) 
                         if ch in region_channels]
            if ch_indices:
                features.append(np.nanmean(sampen_per_channel[ch_indices]))
            else:
                features.append(np.nan)
        
        # Global entropy
        features.append(np.nanmean(sampen_per_channel))
        
        return np.array(features, dtype=np.float64)
    
    def _find_channel(self, name: str, channel_names: list):
        """
        Return the index for a channel name, ignoring capitalization.

        This helper allows feature extraction to work with slightly different
        channel-label casing across datasets.
        """
        for i, ch in enumerate(channel_names):
            if ch.lower() == name.lower():
                return i
        return None
    
    def extract_from_features_csv(self, row: np.ndarray) -> np.ndarray:
        """
        Return a precomputed feature row without recalculating anything.

        This is used when the dashboard reads feature values directly from the
        Sheffield CSV instead of processing raw EEG epochs.
        """
        return row


# GWO-Selected features (13 features from thesis Table 8)
GWO_SELECTED_FEATURES = [
    'rel_theta_frontal',
    'rel_delta_parietal',
    'rel_theta_parietal',
    'abs_theta_occipital',
    'abs_alpha_occipital',
    'theta_alpha_F',
    'coh_delta_C3_C4',
    'coh_alpha_C3_C4',
    'coh_theta_P5_P6',
    'coh_delta_Fz_Pz',
    'coh_beta_F4_P4',
    'sampen_central',
    'sampen_global'
]

File: featureextraction.m
% =========================================================================
%  feature_extraction_1.m
%  EEG Feature Extraction — Sheffield ASD Dataset
%
%  Features extracted per subject:
%   A) Spectral Power (Welch PSD)
%      - Absolute & relative band power (delta/theta/alpha/beta/gamma)
%        per region (frontal/central/parietal/occipital)  → 40 features
%      - Frontal theta/alpha and theta/beta ratios          →  2 features
%
%   B) Functional Connectivity (mscohere)
%      - Mean coherence in 4 bands for 8 electrode pairs   → 32 features
%
%   C) Nonlinear Complexity (Sample Entropy)
%      - Regional average SampEn (4 regions) + global avg  →  5 features
%
%  Total: 79 features
%
%  NOTES:
%   - All channels are guaranteed present after preprocessing (interpolated).
%   - PSD uses Welch with a 1-second Hamming window and 50% overlap.
%   - Coherence uses the same Welch parameters via mscohere.
%   - SampEn uses m=2, r=0.2*std (standard parameters).
%   - Features are averaged across epochs before export.
% =========================================================================

%% ---- Configuration ----
inputDir  = '/Users/admin/Desktop/Sheffield/PROCESSED_V3';
outputCsv = '/Users/admin/Desktop/Sheffield/features_sheffield_noica_withp34.csv';

% Frequency bands
bands.names  = {'delta','theta','alpha','beta','gamma'};
bands.ranges = [1 4; 4 8; 8 15; 15 30; 30 40];

% Region definitions (10-20 labels)
regionDefs.frontal  = {'Fp1','AF7','AF3','F1','F3','F5','F7', ...
                        'Fpz','Fp2','AF8','Fz','F4','F6','F8'};
regionDefs.central  = {'FC5','FC3','FC1','FC4','FC2','C3','C5','C4','C6'};
regionDefs.parietal = {'P1','P3','P5','P7','P9','Pz','P2','P4','P6','P8','P10', ...
                        'CP3','CP4','CP6'};
regionDefs.occipital= {'PO7','PO3','POz','PO8','PO4','O1','Oz','O2','Iz'};

% Coherence pairs
connPairs = { ...
    'F3','F4';   % frontal interhemispheric
    'C3','C4';   % central interhemispheric
    'P5','P6';   % parietal interhemispheric
    'TP7','TP8'; % temporal interhemispheric
    'O1','O2';   % occipital interhemispheric
    'Fz','Pz';   % midline fronto-posterior
    'F3','P5';   % left fronto-posterior
    'F4','P4'};  % right fronto-posterior

connBandNames  = {'delta','theta','alpha','beta'};
connBandRanges = [1 4; 4 8; 8 15; 15 30];

% Sample Entropy parameters
sampEn_m       = 2;    % embedding dimension
sampEn_rFactor = 0.2;  % tolerance = rFactor * std(signal)

%% ---- File list ----
files = dir(fullfile(inputDir, '*.set'));
nSubs = numel(files);
if nSubs == 0
    error('No .set files found in %s', inputDir);
end
fprintf('Found %d preprocessed .set files.\n', nSubs);

%% ---- Build header ----
header       = {'Subject'};
regionNames  = fieldnames(regionDefs);
bandNames    = bands.names;
nBands       = numel(bandNames);
nRegions     = numel(regionNames);

% A) Spectral power
for r = 1:nRegions
    rn = regionNames{r};
    for b = 1:nBands
        bn = bandNames{b};
        header{end+1} = sprintf('abs_%s_%s', bn, rn);  %#ok<SAGROW>
        header{end+1} = sprintf('rel_%s_%s', bn, rn);  %#ok<SAGROW>
    end
end
% Frontal ratios
header{end+1} = 'theta_alpha_F';
header{end+1} = 'theta_beta_F';

% B) Coherence
for p = 1:size(connPairs,1)
    pairName = sprintf('%s_%s', connPairs{p,1}, connPairs{p,2});
    for cb = 1:numel(connBandNames)
        header{end+1} = sprintf('coh_%s_%s', connBandNames{cb}, pairName); %#ok<SAGROW>
    end
end

% C) Sample Entropy
for r = 1:nRegions
    header{end+1} = sprintf('sampen_%s', regionNames{r}); %#ok<SAGROW>
end
header{end+1} = 'sampen_global';

nFeat    = numel(header) - 1;  % exclude Subject column
featMat  = nan(nSubs, nFeat);
subNames = cell(nSubs, 1);

fprintf('Feature vector size: %d\n', nFeat);

%% ---- Main loop ----
for s = 1:nSubs
    fname = files(s).name;
    fpath = fullfile(inputDir, fname);
    fprintf('\n[%d/%d] %s\n', s, nSubs, fname);

    try
        %% Load
        EEG = pop_loadset('filename', fname, 'filepath', inputDir);
        EEG = eeg_checkset(EEG);

        data   = EEG.data;                  % ch × samples × epochs
        fs     = EEG.srate;
        [nCh, nSamp, nEp] = size(data);
        labels = {EEG.chanlocs.labels};
        fprintf('  %d ch | %d samples/epoch | %d epochs | fs=%.0f Hz\n', ...
            nCh, nSamp, nEp, fs);

        %% Welch parameters (1-second window, 50% overlap)
        winLen   = min(fs, nSamp);  % 256 samples = 1 s (or shorter if epoch < 1 s)
        win      = hamming(winLen);
        noverlap = floor(winLen / 2);
        freqRes  = 0.5;             % Hz resolution
        freqVec  = 0:freqRes:40;    % frequencies at which PSD is evaluated

        %% ===== A) Spectral features =====
        % absPow: nEp × nCh × nBands
        absPow = nan(nEp, nCh, nBands);

        for ch = 1:nCh
            for ep = 1:nEp
                sig = double(squeeze(data(ch,:,ep)));
                if all(sig == 0) || any(isnan(sig)); continue; end

                [Pxx, f] = pwelch(sig, win, noverlap, freqVec, fs);

                for b = 1:nBands
                    fr  = bands.ranges(b,:);
                    idx = f >= fr(1) & f <= fr(2);
                    if any(idx)
                        absPow(ep,ch,b) = trapz(f(idx), Pxx(idx));
                    end
                end
            end
        end

        % Average over epochs → nCh × nBands
        absPow_ch = squeeze(mean(absPow, 1, 'omitnan'));

        % Total broadband power per channel (sum across bands)
        totalPow_ch = sum(absPow_ch, 2);                     % nCh × 1
        totalPow_ch(totalPow_ch == 0) = NaN;                 % avoid /0

        % Relative power
        relPow_ch = absPow_ch ./ totalPow_ch;                % nCh × nBands

        % Regional averages
        regionBands_abs = nan(nRegions, nBands);
        regionBands_rel = nan(nRegions, nBands);

        for r = 1:nRegions
            rn      = regionNames{r};
            chanList= regionDefs.(rn);
            idxR    = find(ismember(labels, chanList));

            if isempty(idxR)
                warning('  No channels for region %s — row will be NaN.', rn);
                continue;
            end
            regionBands_abs(r,:) = mean(absPow_ch(idxR,:), 1, 'omitnan');
            regionBands_rel(r,:) = mean(relPow_ch(idxR,:), 1, 'omitnan');
        end

        % Frontal theta/alpha and theta/beta ratios
        iF     = find(strcmp(regionNames, 'frontal'));
        iTheta = find(strcmp(bandNames,  'theta'));
        iAlpha = find(strcmp(bandNames,  'alpha'));
        iBeta  = find(strcmp(bandNames,  'beta'));

        if ~isempty(iF) && ~isempty(iTheta) && ~isempty(iAlpha) && ~isempty(iBeta)
            thetaF = regionBands_rel(iF, iTheta);
            alphaF = regionBands_rel(iF, iAlpha);
            betaF  = regionBands_rel(iF, iBeta);
            thetaAlphaF = thetaF / alphaF;
            thetaBetaF  = thetaF / betaF;
        else
            thetaAlphaF = NaN;
            thetaBetaF  = NaN;
        end

        %% ===== B) Coherence features =====
        nPairs     = size(connPairs,1);
        nConnBands = size(connBandRanges,1);
        cohFeat    = nan(nPairs, nConnBands);

        for p = 1:nPairs
            ch1   = connPairs{p,1};
            ch2   = connPairs{p,2};
            idx1  = find(strcmpi(labels, ch1));
            idx2  = find(strcmpi(labels, ch2));

            if isempty(idx1) || isempty(idx2)
                fprintf('  Pair %s-%s missing — coherence = NaN.\n', ch1, ch2);
                continue;
            end

            coh_ep = nan(nEp, nConnBands);

            for ep = 1:nEp
                sig1 = double(squeeze(data(idx1,:,ep)));
                sig2 = double(squeeze(data(idx2,:,ep)));

                if all(sig1==0)||all(sig2==0)||any(isnan(sig1))||any(isnan(sig2))
                    continue;
                end

                [Cxy, fC] = mscohere(sig1, sig2, win, noverlap, freqVec, fs);

                for cb = 1:nConnBands
                    fr    = connBandRanges(cb,:);
                    idxF  = fC >= fr(1) & fC <= fr(2);
                    if any(idxF)
                        coh_ep(ep,cb) = mean(Cxy(idxF));
                    end
                end
            end

            cohFeat(p,:) = mean(coh_ep, 1, 'omitnan');
        end

        %% ===== C) Sample Entropy =====
        sampEn_ch_ep = nan(nCh, nEp);

        for ch = 1:nCh
            for ep = 1:nEp
                sig = double(squeeze(data(ch,:,ep)));
                if all(sig==0)||any(isnan(sig)); continue; end
                r_tol = sampEn_rFactor * std(sig);
                if r_tol == 0; continue; end
                sampEn_ch_ep(ch,ep) = SampEn_1D(sampEn_m, r_tol, sig);
            end
        end

        sampEn_ch = mean(sampEn_ch_ep, 2, 'omitnan');  % nCh × 1

        sampEn_region = nan(nRegions,1);
        for r = 1:nRegions
            rn      = regionNames{r};
            chanList= regionDefs.(rn);
            idxR    = find(ismember(labels, chanList));
            if isempty(idxR); continue; end
            sampEn_region(r) = mean(sampEn_ch(idxR), 'omitnan');
        end

        sampEn_global = mean(sampEn_ch, 'omitnan');

        %% ===== Assemble feature row =====
        featRow = nan(1, nFeat);
        k = 1;

        % A) Regional abs & rel power (region × band × 2)
        for r = 1:nRegions
            for b = 1:nBands
                featRow(k) = regionBands_abs(r,b); k = k+1;
                featRow(k) = regionBands_rel(r,b); k = k+1;
            end
        end

        % Frontal ratios
        featRow(k) = thetaAlphaF; k = k+1;
        featRow(k) = thetaBetaF;  k = k+1;

        % B) Coherence
        for p = 1:nPairs
            for cb = 1:nConnBands
                featRow(k) = cohFeat(p,cb); k = k+1;
            end
        end

        % C) Sample Entropy
        for r = 1:nRegions
            featRow(k) = sampEn_region(r); k = k+1;
        end
        featRow(k) = sampEn_global; k = k+1; %#ok<NASGU>

        %% Sanity check
        if k-1 ~= nFeat
            error('Feature count mismatch: expected %d, got %d', nFeat, k-1);
        end

        featMat(s,:) = featRow;
        subNames{s}  = fname;
        fprintf('  Done. NaN features: %d/%d\n', sum(isnan(featRow)), nFeat);

    catch ME
        fprintf('  ERROR: %s\n', ME.message);
        subNames{s} = fname;
    end
end

%% ---- Write CSV ----
T = array2table(featMat, 'VariableNames', header(2:end));
T.Subject = subNames;
T = movevars(T, 'Subject', 'Before', 1);

fprintf('\nWriting to %s ...\n', outputCsv);
writetable(T, outputCsv);
fprintf('Done. %d subjects × %d features.\n', nSubs, nFeat);

%% ============================================================
%  Helper: Sample Entropy (1D, Chebyshev distance)
% ============================================================
function se = SampEn_1D(m, r, x)
% SampEn_1D  Compute sample entropy of a 1-D time series.
%
%   m  : embedding dimension
%   r  : tolerance (absolute, typically 0.2*std(x))
%   x  : input vector
%
%   Reference: Richman & Moorman, Am J Physiol 2000.

x = x(:)';
N = length(x);

if N <= m+2
    se = NaN;
    return;
end

% Pre-allocate template matrices
Xm  = zeros(N-m,   m  );
Xm1 = zeros(N-m-1, m+1);

for ii = 1:(N-m)
    Xm(ii,:)  = x(ii:ii+m-1);
end
for ii = 1:(N-m-1)
    Xm1(ii,:) = x(ii:ii+m);
end

% Count template matches (excluding self-matches, consistent with SampEn)
count_m  = 0;
count_m1 = 0;

for ii = 1:size(Xm,1)-1
    dist_m  = max(abs(Xm(ii+1:end,:)  - Xm(ii,:)),  [], 2);
    count_m = count_m + sum(dist_m <= r);

    dist_m1  = max(abs(Xm1(ii+1:end,:) - Xm1(ii,:)), [], 2);
    count_m1 = count_m1 + sum(dist_m1 <= r);

end

if count_m == 0 || count_m1 == 0
    se = NaN;  % undefined (no matches found at this tolerance)
else
    se = -log(count_m1 / count_m);
end
end
